### PR TITLE
    feat(tianmu): Supports user-defined functions (#538)

### LINKED ISSUE #538: feature: The Tianmu engine supports user  defined function
### Abstract:

The current column storage engine does not support custom functions. Previously, it used the Mysql/SQL layer to parse custom functions. But this led to a series of data errors.
This enables formal column storage engine-based support for custom functions.
This paper analyzes its requirements.

### Functional Requirements:

1. The result should be consistent with the result of InnoDB calling custom function to ensure the correctness of the result
2. Do not use callback to the SQL layer to parse

The callback to the SQL layer will result in the coupling between the engine layer and the SQL layer, and the logical flow cannot be clearly understood

If the engine layer and SQL layer are tightly coupled, it is difficult to upgrade the mysql version

Each upgrade requires writing a call to the engine layer based on the implementation of the new mysql SQL layer

Each time, the call to the engine is rewritten, increasing the workload and increasing the time it takes to test the code

### Custom functions only support mysql scalar data, not array or composite data types

1. Compatible with mysql custom function scalar usage

2. The return value of the function does not support json type and array compound type

 

#### 1. the time required to call custom functions

Not planned, not considered

#### 2. Upper limit of memory usage

Not planned, not considered

#### 3. Upper limits on CPU usage

Not planned, not considered

### LINKED ISSUE #851: bug: ERROR 1436 (HY000): Thread stack overflow
### Have you read the Contributing Guidelines on issues?

- [X] I have read the [Contributing Guidelines on issues](https://github.com/stoneatom/stonedb/blob/stonedb-5.7-dev/CONTRIBUTING.md).

### Please confirm if bug report does NOT exists already ?

- [X] I confirm there is no existing issue for this

### Context of scene

https://adofsauron.blog.csdn.net/article/details/126788075

### The thread stack overflow occurred. Procedure

```sql

ERROR 1436 (HY000): Thread stack overrun:  11370160688 bytes used of a 524288 byte stack, and 128000 bytes needed.  Use 'mysqld --thread_stack=#' to specify a bigger stack.


```

### DML

```sql

CREATE TABLE `CUSTOM` (
  `USID` varchar(50) NOT NULL COMMENT '????',
  `PASSWORD` varchar(32) NOT NULL COMMENT '????',
  `LGTP` varchar(2) NOT NULL COMMENT '????',
  `USTP` varchar(2) NOT NULL COMMENT '????',
  `STATUS` varchar(2) NOT NULL COMMENT '??????',
  `USQX` varchar(20) NOT NULL COMMENT '????',
  `AUTOFAPIAO` int(11) NOT NULL COMMENT '????????????',
  `USDJ` int(11) DEFAULT NULL COMMENT '????',
  `ISZHIFUSELF` int(11) DEFAULT NULL COMMENT '?????????',
  `ZHIFUID` varchar(50) DEFAULT NULL COMMENT '????????',
  `ISAUDITSELF` int(11) DEFAULT NULL COMMENT '?????????',
  `AUDUSID` varchar(50) DEFAULT NULL COMMENT '????????',
  `TTLB` varchar(2) DEFAULT NULL COMMENT '????',
  `SUSID` varchar(50) DEFAULT NULL COMMENT '?????????',
  `CORPNAME` varchar(100) DEFAULT NULL COMMENT '????',
  `TOURLICENSECODE` varchar(50) DEFAULT NULL COMMENT '???????',
  `BUSINESSLICENSESCODE` varchar(50) DEFAULT NULL COMMENT '??????',
  `TOURLICENSEUPID` int(11) DEFAULT NULL COMMENT '????????',
  `BUSINESSLICENSESUPID` int(11) DEFAULT NULL COMMENT '??????',
  `LNAME` varchar(50) DEFAULT NULL COMMENT '?????',
  `ZJLB` varchar(2) DEFAULT NULL COMMENT '???????',
  `ZJHM` varchar(50) DEFAULT NULL COMMENT '????',
  `IREGIONALID` int(11) DEFAULT NULL COMMENT '???',
  `BANK` varchar(100) DEFAULT NULL COMMENT '????',
  `ACCOUNT` varchar(50) DEFAULT NULL COMMENT '????',
  `BNAME` varchar(100) DEFAULT NULL COMMENT '???',
  `ADDR` varchar(200) DEFAULT NULL COMMENT '??',
  `TELNO` varchar(50) DEFAULT NULL COMMENT '????',
  `MOBILE` varchar(50) DEFAULT NULL COMMENT '????',
  `FAXNO` varchar(50) DEFAULT NULL COMMENT '????',
  `QQ` varchar(50) DEFAULT NULL COMMENT 'QQ??',
  `MSN` varchar(50) DEFAULT NULL COMMENT 'MSN??',
  `POSTNO` varchar(6) DEFAULT NULL COMMENT '????',
  `EMAIL` varchar(50) DEFAULT NULL COMMENT 'email??',
  `LDATE` varchar(20) NOT NULL COMMENT '????',
  `LASTDATE` varchar(20) DEFAULT NULL COMMENT '??????',
  `LOGTIMES` varchar(50) DEFAULT NULL COMMENT '??????',
  `NOTEA` varchar(100) DEFAULT NULL,
  `NOTEB` text,
  `TIMES` int(11) DEFAULT NULL COMMENT '??????',
  `LMDATE` varchar(20) DEFAULT NULL COMMENT '??????????',
  `LPDATE` varchar(20) NOT NULL COMMENT '????????',
  `CDATE` varchar(20) DEFAULT NULL COMMENT '??????',
  `MMQT` varchar(4) DEFAULT NULL COMMENT '??????',
  `MMDA` varchar(100) DEFAULT NULL COMMENT '??????',
  `NOTE1` varchar(100) DEFAULT NULL,
  `NOTE2` varchar(100) DEFAULT NULL,
  `NOTE3` varchar(100) DEFAULT NULL,
  `NOTE4` varchar(100) DEFAULT NULL,
  `NOTE5` varchar(100) DEFAULT NULL,
  `NOTE6` varchar(100) DEFAULT NULL,
  `NOTE7` varchar(100) DEFAULT NULL,
  `NOTE8` varchar(100) DEFAULT NULL,
  `NOTE9` varchar(100) DEFAULT NULL,
  `NOTE10` varchar(100) DEFAULT NULL,
  `INOTE1` int(11) DEFAULT NULL,
  `INOTE2` int(11) DEFAULT NULL,
  `INOTE3` int(11) DEFAULT NULL,
  `INOTE4` int(11) DEFAULT NULL,
  `INOTE5` int(11) DEFAULT NULL,
  `INOTE6` int(11) DEFAULT NULL,
  `INOTE7` int(11) DEFAULT NULL,
  `INOTE8` int(11) DEFAULT NULL,
  `INOTE9` int(11) DEFAULT NULL,
  `INOTE10` int(11) DEFAULT NULL,
  `IBUSINESSID` int(11) NOT NULL COMMENT '??ID',
  `DAOYOUNO` varchar(50) DEFAULT NULL COMMENT '????',
  `DTMAKEDATE` datetime DEFAULT CURRENT_TIMESTAMP ON UPDATE CURRENT_TIMESTAMP,
  `PAYPASSWORD` varchar(32) DEFAULT NULL,
  `AGENTID` decimal(18,0) DEFAULT NULL,
  `OPENUSERID` varchar(100) DEFAULT NULL,
  `OPENID` varchar(100) DEFAULT NULL,
  `QRCODEURL` varchar(100) DEFAULT NULL,
  `PROMOTERID` varchar(100) DEFAULT NULL,
  PRIMARY KEY (`USID`) USING BTREE
) ENGINE=TIANMU DEFAULT CHARSET=utf8;

```

### user defined function

```sql

    DELIMITER ;;
    CREATE DEFINER=`root`@`%` FUNCTION `FIND_TOP_USID`(PUSID VARCHAR(1000)) RETURNS varchar(1000) CHARSET utf8
    BEGIN
    DECLARE FINALVAR VARCHAR(100);
    WHILE LENGTH(PUSID) > 0 DO 
    SET FINALVAR = PUSID;
    SELECT SUSID INTO PUSID FROM CUSTOM WHERE USID = FINALVAR;
    IF FINALVAR = PUSID 
    THEN SET PUSID = ''; 
    END IF;
    END WHILE;
    RETURN FINALVAR;
    END ;;
    DELIMITER ;

```

### Query SQL

```sql

    SELECT sale.usid AS usid,
        cus.corpname AS corpname,
        sale.iscenicid AS iscenicid,
        salde.itickettypeid AS itickettypeid,
        price.icrowdkindid AS icrowdkindid,
        salde.mactualsaleprice AS mactualsaleprice,
        sum(salde.iamount) AS numb,
        sum(salde.meventmoney) AS mont,
        coalesce(sum(salde.mderatemoney), 0) AS mderatemoney,
        coalesce(sum(salde.ideratenums), 0) AS ideratenums,
        coalesce(sum(salde.mhandcharge), 0) AS mhandcharge,
        saletype.isettlementid AS isettlementid,
        sale.bysalesvouchertype AS bysalesvouchertype
    FROM STSSALESVOUCHERTAB sale,
        STSSALESVOUCHERDETAILSTAB salde,
        (SELECT C.SUSID,
                C2.CORPNAME AS SCORPNAME,
                C2.BNAME,
                C.USID,
                C.CORPNAME,
                C.IBUSINESSID,
                C.TTLB
            FROM (SELECT USID,
                        CORPNAME,
                        IBUSINESSID,
                        TTLB,
                        FIND_TOP_USID(USID) AS SUSID
                    FROM CUSTOM
                    WHERE LGTP = '02'
                    AND USTP = '01'
                    AND IBUSINESSID = 2) C
            LEFT JOIN CUSTOM C2
                ON C2.USID = C.SUSID) cus,
        EDMTICKETTYPETAB ti,
        EDMCROWDKINDPRICETAB price,
        STSSALESSETTLEMENTTAB saletype
    WHERE substr(sale.dtmakedate, 1, 10) >= '2022-08-18'
    AND substr(sale.dtmakedate, 1, 10) <= '2022-08-18'
    AND sale.isalesvoucherid = saletype.isalesvoucherid
    AND sale.iticketstationid = saletype.iticketstationid
    AND price.icrowdkindpriceid = salde.icrowdkindpriceid
    AND sale.isalesvoucherid = salde.isalesvoucherid
    AND sale.iticketstationid = salde.iticketstationid
    AND cus.usid = sale.usid
    AND salde.itickettypeid = ti.itickettypeid
    AND sale.iscenicid IN (1)
    AND cus.ibusinessid <> 1
    GROUP BY sale.usid,
            cus.corpname,
            saletype.isettlementid,
            salde.mactualsaleprice,
            sale.iscenicid,
            salde.itickettypeid,
            price.icrowdkindid,
            sale.bysalesvouchertype
    UNION ALL
    SELECT sale.usid AS usid,
        cus.corpname AS corpname,
        sale.iscenicid AS iscenicid,
        salde.itickettypeid AS itickettypeid,
        price.icrowdkindid AS icrowdkindid,
        salde.mactualsaleprice AS mactualsaleprice,
        sum(salde.iamount) AS numb,
        sum(salde.meventmoney) AS mont,
        coalesce(sum(salde.mderatemoney), 0) AS mderatemoney,
        coalesce(sum(salde.ideratenums), 0) AS ideratenums,
        coalesce(sum(salde.mhandcharge), 0) AS mhandcharge,
        saletype.isettlementid AS isettlementid,
        sale.bysalesvouchertype AS bysalesvouchertype
    FROM STSSALESVOUCHERTABLIST sale,
        STSSALESVOUCHERDETAILSTABLIST salde,
        (SELECT USID,
                CORPNAME,
                IBUSINESSID,
                TTLB,
                FIND_TOP_USID(USID) AS SUSID
            FROM CUSTOM
            WHERE LGTP = '02'
            AND USTP = '01') cus,
        EDMTICKETTYPETAB ti,
        EDMCROWDKINDPRICETAB price,
        STSSALESSETTLEMENTTABLIST saletype
    WHERE substr(sale.dtmakedate, 1, 10) >= '2022-08-18'
    AND substr(sale.dtmakedate, 1, 10) <= '2022-08-18'
    AND sale.isalesvoucherid = saletype.isalesvoucherid
    AND sale.iticketstationid = saletype.iticketstationid
    AND price.icrowdkindpriceid = salde.icrowdkindpriceid
    AND sale.isalesvoucherid = salde.isalesvoucherid
    AND sale.iticketstationid = salde.iticketstationid
    AND cus.usid = sale.usid
    AND salde.itickettypeid = ti.itickettypeid
    AND sale.iscenicid IN (1)
    AND cus.ibusinessid <> 1
    GROUP BY sale.usid,
            cus.susid,
            saletype.isettlementid,
            salde.mactualsaleprice,
            sale.iscenicid,
            salde.itickettypeid,
            price.icrowdkindid,
            sale.bysalesvouchertype;

```

### Expected behavior

_No response_

### How To Reproduce

_No response_

### Environment

_No response_

### Are you interested in submitting a PR to solve the problem?

- [ ] Yes, I will!

### PR DESCRIPTION
**Pull request by adofsauron**
Requirement analysis:
    https://adofsauron.blog.csdn.net/article/details/126991367

    Changes description:
    https://blog.csdn.net/adofsauron/article/details/127543458

## Summary about this PR

Issue Number: close #538 

## Tests Check List

- [ ] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

## Changelog

- [x] New Feature
- [ ] Bug Fix
- [ ] Improvement
- [ ] Performance Improvement
- [ ] Build/Testing/CI/CD
- [ ] Documentation
- [ ] Not for changelog (changelog entry is not required)

## Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features

### CODE REVIEW

File: mysql-test/suite/tianmu/r/custom_function.result [hunk 1]
```diff
@@ -0,0 +1,9 @@
+use test;
```
Comment (RingsC) [inline mysql-test/suite/tianmu/r/custom_function.result]: using "--source have_tianmu.inc"

Comment (RingsC) [reply mysql-test/suite/tianmu/r/custom_function.result]: using --disable_query_log/ --enable_query_log to shut down the unused log.

Comment (RingsC) [reply mysql-test/suite/tianmu/r/custom_function.result]: the file name should be attached to an issue.

**Review by RingsC — CHANGES_REQUESTED** [review]

File: mysql-test/suite/tianmu/t/issue538.test [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+  `employee_id` int(11) NOT NULL ,
+  `employee_name` varchar(50) NOT NULL,
+  `employee_sex` varchar(10) DEFAULT 'men',
+  `hire_date` datetime NOT NULL DEFAULT CURRENT_TIMESTAMP,
+  `employee_mgr` int(11) DEFAULT NULL,
+  `employee_salary` float DEFAULT '3000',
+  `department_id` int(11) DEFAULT NULL
+) ENGINE=TIANMU DEFAULT CHARSET=utf8mb4;
+
+--disable_warnings
```
Comment (RingsC) [inline mysql-test/suite/tianmu/t/issue538.test]: Why do u use disable warnings twice?

File: mysql-test/suite/tianmu/t/issue538.test [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+SELECT
+	a.employee_id,
+	get_value(b.employee_id)
+FROM
+	employees a
+LEFT JOIN employees b ON
+	a.employee_id = b.employee_id
+	and b.employee_name = 'David Tian';
+
+SELECT get_value(1);
```
Comment (RingsC) [inline mysql-test/suite/tianmu/t/issue538.test]: I think we need these test.  Taking an instance: 
select sum (xx + get_value(1)) xxx
or 
select xxx from xxx where xxx oper get_value. 
....
etc.

**Review by RingsC — CHANGES_REQUESTED** [review]

File: mysql-test/suite/tianmu/t/issue538.test (hunk 3 above)
Comment (adofsauron) [reply mysql-test/suite/tianmu/t/issue538.test]: sure, wait a time for add this

**Pull request closed without merging**